### DISCUSSION TITLE
blog/angular-object-inputs/

### DISCUSSION THREAD

**Discussion by giscus** [nartc/tnc-scully#54] [Announcements]
# blog/angular-object-inputs/

Personal blog by Chau Tran

https://nartc.me/blog/angular-object-inputs/

**Comment by khangtrannn** [comment]
Thanks anh Châu, this article is exactly what I need to solve the problem that I recently experienced 🤩

**Comment by Khaldor48** [comment]
Great idea, Chau strikes again :) thank you

**Comment by Alex88360** [comment]
That doesn't work because typescript can't assure that the merged object will conform to the TOptions type (readonly property, discriminated unions).

**Reply by nartc** [reply]
Thanks for the comment. Yeah, the merge here is idiomatic. However for discriminated unions, not sure how it affects the consumers though since you’d still required to pass in the proper union member. 

re:readonly property
I’d love to see an example where this is an issue for the consumers too.

**Comment by DaniKurosaki** [comment]
Hi Chau, nice post!
Wanted to ask a couple doubts: I guess this Object Input approach is suited for Dumb Components but not for Smart Components. Since Smart change some values of said object, how would you approach this? Making the Object Input a ModelSignal instead of an InputSignal seems overkill IMO, since you will have updating access to variables you should not be able to update

**Reply by nartc** [reply]
Thanks.

I think `linkedSignal` would help but if you have a use-case, I'd love to see it. When I hear "smart components that can _control_ their internal state", `linkedSignal` + `input` come to mind rather than `model` input.

If a 2-way data binding is needed, I think separating some inputs out into `model` inputs and keep the rest in the Object Input would make more sense than making the entire Object Input as `model`

For example in the post, let's say we can do `[(isPaused)]` and the Physics component can also control whether to pause the physics simulation. Then I would make `isPaused = model(false)` on its own

**Reply by DaniKurosaki** [reply]
Here is one example, not a complex one but its the first one that came to mind:

```typescript
@Component({
	selector: "progress-slider",
	standalone: true,
	imports: [],
	templateUrl: "./progress-slider.component.html",
	styleUrls: ["./progress-slider.component.scss"],
	changeDetection: ChangeDetectionStrategy.OnPush,
	host: {
		"[class.clickable]": "clickable$()",
		"(mousedown)": "startSlide($event)",
		"(mousemove)": "moveSlide($event)",
		"(mouseup)": "stopSlide($event)",
	},
})
export class ProgressSliderComponent {
	/**
	 * ElementRef instance
	 */
	private readonly elementRef: ElementRef = inject(ElementRef);

	/**
	 * Current value
	 */
	public current$: ModelSignal<number> = model.required<number>();

	/**
	 * Maximum value
	 */
	public max$: InputSignal<number> = input.required<number>();

	/**
	 * Maximum value
	 */
	public minValue$: InputSignal<number> = input<number>(0);

	/**
	 * Slider's label, displayed if provided
	 */
	public label$: InputSignal<string | undefined> = input<string>();

	/**
	 * Indicates if the slider is clickable, which means the user can change the value by clicking on it. Applies the "clickable" class to the host
	 */
	public clickable$: InputSignal<boolean> = input<boolean>(true);

	/**
	 * Indicates if the slider is sliding, which means the user is dragging the slider
	 */
	public isSliding$: WritableSignal<boolean> = signal(false);

	/**
	 * Host's native element
	 */
	public get hostElement(): HTMLElement {
		return this.elementRef.nativeElement;
	}

	/**
	 * Starts the sliding process
	 * @param { MouseEvent } event mouse event
	 */
	startSlide(event: MouseEvent): void {
		if (!this.clickable$()) return;
		event.preventDefault();
		event.stopPropagation();

		this.isSliding$.set(true);
		this.moveSlide(event);
	}

	/**
	 * Listens to a mousemove event to move the slider. It only works if the slider is clickable and is sliding
	 * @param { MouseEvent } event mouse event
	 */
	moveSlide(event: MouseEvent): void {
		if (!this.clickable$() || !this.isSliding$()) return;
		event.preventDefault();
		event.stopPropagation();

		const rect = this.hostElement.getBoundingClientRect();
		let progressPosition = event.clientX - rect.left;

		if (progressPosition < 0) progressPosition = 0;
		if (progressPosition > rect.width) progressPosition = rect.width;

		let newValue = Math.round((progressPosition / rect.width) * this.max$());

		const minValue = this.minValue$();
		if (newValue < minValue) newValue = minValue;

		this.current$.set(newValue); // Aquí obtienes el valor en porcentaje, puedes adaptarlo según tus necesidades.
	}

	/**
	 * Stops the sliding process
	 */
	stopSlide(event: MouseEvent): void {
		if (!this.clickable$()) return;
		event.preventDefault();
		event.stopPropagation();

		this.isSliding$.set(false);
	}
}
```
